API for value updates (#3532)
enso-org/enso: docs/language-server/protocol-language-server.md

```typescript
interface TextEdit {
  /** The range of text in a text file. */
  range: Range;

  /** The change to a text file. */
  text: String;

  /** The update type. When not set defaults to `FileEditKind`. */
  kind?: EditKind;
}
```

> ### EditKind

Reviewer: I would suggest that instead of adding a `EditKind` we should have two flags, that specify the follow-up behaviour of the edit.
`recompile: bool` and `re-execute: bool`. 
This would mean that it is the IDEs responsibility to decide which of these is appropriate for a given edit or sequence of edits. The docs should mention the intended use case (.e.g., edits to metadata). 

This avoids inconsistent edits, where the `EditKind` is incompatible with the actual edit. Presently, the API happily accepts an update to any part of thee file labelled as `MetadataEditKind`, which doesn't seem to make sense. 
While of course someone can still set the flags wrong, the API would be more consistent in itself.

For the same reason, I would suggest creating a new API endpoint for `ValueUpdates`, that also does not allow inconstant data to be provided.

Something like 
```
type ExpressionUpdate {
    expressionId: ExpressionId;
    new_expression: String;
}
```
This update would always be consistent.

Author: If IDE can take control of the execution, then the new endpoint makes sense. The one downside is that with this design IDE needs to make two requests for every slider update instead of one - one editing the expression, and second setting the value. Do you think this tradeoff is ok?

Reviewer: > The one downside is that with this design IDE needs to make two requests for every slider update instead of one - one editing the expression, and second setting the value.

Can you give an example? I might be missing something here.

Reviewer: > If IDE can take control of the execution, then the new endpoint makes sense. 

My main concern is, that the current proposal allows for nonsense to be submitted. In a perfect world, I guess the engine should be able to detect whether a change requires what kinds of follow-up action, but without that, I think we are better off to have a more explicit API.

Author: > Can you give an example? I might be missing something here.

For example, we have a slider on node allowing to change the value in the range [0, 100]
- User moves the slider from 18 to 19
- IDE sends `applyEdit execute=false` request to change the program text
- IDE sends `setExpression` request to set the node value to 19

Reviewer 2: Right. `setExpression` changes not only the value, but also the underlying text. It is desirable to keep text and AST consistent. There are three options:
- combined `applyEdit` message where IDE requests both - the text change as well as _hints_ the _literal change_
- two messages from the IDE `applyEdit execute=false` and `setExpression`
-  one message from IDE to `setExpression` and _one message from the engine_ to the IDE that the code has been modified

There are various pros/cons to each option. Some favor consistency, some favor correctness.

Author: > one message from IDE to setExpression and one message from the engine to the IDE that the code has been modified

I like this one. We can even skip the Engine->IDE notification assuming that `setExpression` always changes the source.
The downside is that we can lose consistency because we don't send `SHA` before the change and `SHA` after, like we do in `applyEdit`.
Although we can add `oldVersion` and `newVersion` with `SHA` hashes to the `setExpression`.
But assuming that we want to enable the optimization with sending metadata, I'd propose the following:

Add `execute` flag to the `text/applyEdit` message indicating whether or not the program should be re-executed after applying the change.
```typescript
interface ApplyEdit {
  edit: FileEdit;
  /** New flag controlling the program execution. `true` if not set (current behavior). */
  execute?: boolean;
}
```
Create the new command `text/applyExpressionValue` that **sets** the new expression value *and* **modifies** the source code.
```typescript
interface ApplyExpressionValue {
  /** Edited expression id. */
  expressionId: ExpressionId;
  /** The new value of the node. */
  expressionValue: string;
  /** Edited file. */
  path: Path;
  /** The sha hash of the file before setting the the new value to the node. */
  oldVersion: SHA3-224;
  /** The sha hash of the file after setting the the new value to the node. */
  newVersion: SHA3-224;
}

Represents the type of an update enabling runtime optimizations.

- `FileEditKind` represents a plane file update. After applying the text edit,
  the program will be recompiled and re-executed.
- `MetadataEditKind` represents the metadata update. Metadata update does not
  change the program and thus allows engine to skip inane re-executions.
- `ValueUpdateKind` represents an update of a node value. In this case the
  engine can skip the compilation phase, and re-execute the program with a new
  value.

#### Format

```typescript
type EditKind = FileEditKind | MetadataEditKind | ValueUpdateKind;

interface FileEditKind {}

interface MetadataEditKind {}

interface ValueUpdateKind {
  /** The identifier of altered node. */

>   expressionId: ExpressionId;

Reviewer 2: Currently no `expressionId` is needed. What the code will use it for?

Author: IDE knows the id of expression it edits. And for the engine it should be easier to search the node by it's id rather than by location

Reviewer 2: OK. Rewriting the _search for a literal_ in `IR` and Truffle nodes to `expressionId` shall make the system more robust. Let's do it!

  /** The new value of altered node. */

>   value: string;

Reviewer 2: So far the value is taken from `TextEdit.text`. Can you describe what is the relation between `text` and `value` in the documentation? 

Interesting case is: if there is an original value `12345` and the user deletes `3` and gets `1245`. What will the IDE send as a message?

Author: The `value` field covers exactly this case. IDE sends an edit removing the character between positions 2 and 3, and the new value (in pseudo json)
```json
{ 
  "edit" : {
    "range" : [2,3],
    "text" : "",
    "kind" : {
      "expressionId" : "uuid",
      "value" : "1245"
    }
}
```

Reviewer 2: Making the IDE aware of _identifier change_ is a great intention. However when the engine applies the change it also needs to rebuild the `IR` (and Truffle nodes). It is not entirely clear how _correct_ the new IR has to be, but the current implementation needs the new `IR` element to have proper location - see [ChangesetBuilder](https://github.com/enso-org/enso/blob/c72a6582bc62ced2e7da200688728774d661fb7b/engine/runtime/src/main/scala/org/enso/compiler/context/ChangesetBuilder.scala#L90) - otherwise subsequent changes may not find the right element to change.

Right now the beginning of the `IR` is the same as the initial edit location. That's no longer going to be the case if the edit is done _in middle_ of an identifier....

Maybe it is not a problem if `expressionId` is used more extensively. Just make sure subsequent changes to the literal work.

  access to a resource.
- [`FileNotFound`](#filenotfound) informs that file cannot be found.

### `text/openBuffer`

This requests the language server to open a specified in-memory buffer mapped to
the provided path. If the path exists, this command behaves the same as
[`text/openFile`](#textopenfile). If the path does not exist, the command
creates empty in-memory buffer for the provided path.

- **Type:** Request
- **Direction:** Client -> Server
- **Connection:** Protocol
- **Visibility:** Public

If no client has write lock on the opened file, the capability is granted to the

> client that sent the `text/openBuffer` message.

Reviewer: It might be good for the docs to briefly mention how to read from the buffer and what a possible use case for this buffer is.

- [`WriteDeniedError`](#writedeniederror) to signal that the client doesn't hold
  write lock for the buffer.

### `text/applyExpressionValue`

> This requests to apply specific value to the expression. I.e. it can be an
> update of a literal value, like changing `98` to `99`, `true` to `false` or
> `"Hello"` to `"World!"`. This method is a more specific version of

Reviewer: ```suggestion
This requests to set an expression to a new value. For example,
it can update a literal value, like changing `98` to `99`, `true` to `false` or
`"Hello"` to `"World!"`. This method is a more specific version of
```
